### DISCUSSION TITLE
CL code from ChatGPT :)

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#200] [General]
So just for fun, a friend asked ChatGPT to write a CL function that did OpenGL picking. The result is below. I assume GPT found it in a GitHub or such somewhere, but a cursory Google search on the code terms returns no obvious results. I have not tested the code.

```
(defun pick-object (x y)
  "Perform OpenGL picking at the given screen coordinates (x, y)."
  (let* ((viewport (gl:get :viewport))
         (width (first viewport))
         (height (second viewport))
         (window-z (gl:get-pixel (round x) (round y) :depth))
         (modelview-matrix (gl:get :modelview-matrix))
         (projection-matrix (gl:get :projection-matrix))
         (winx (/ (+ x 1.0) (* 2.0 width)))
         (winy (/ (- height y 1.0) (* 2.0 height)))
         (winz window-z)
         (objx (* 2.0 winx -1.0))
         (objy (* 2.0 winy -1.0))
         (objz (* 2.0 winz -1.0))
         (inverse-projection-matrix (invert projection-matrix))
         (inverse-modelview-matrix (invert modelview-matrix))
         (obj-coords (matrix-multiply inverse-projection-matrix
                                      (vec4 objx objy objz 1.0)))
         (obj-coords (matrix-multiply inverse-modelview-matrix
                                      obj-coords)))
    (multiple-value-bind (x y z w) obj-coords
      (vec3 (/ x w) (/ y w) (/ z w)))))
```

**Comment by jolby** [comment]
ChatGPT is pretty crazy. It doesn't produce immediately usable code (at least for me), but it gets eerily close a lot of times. Earlier today I did a similar experiment, with the use-case of writing a SDL2 inner event loop for McCLIM. This first cut wasn't too bad, but then it blew my mind when I asked it to convert the case based event type dispatch to CLOS based dispatch. The result was pretty much perfect! This is the transcript of my ChatGPT conversation:

https://gist.github.com/jolby/e1cfb217a411603ed1ee446d9d483a6c

**Reply by kaveh808** [reply]
Wow, this is pretty amazing. I wonder how much of it is smoke and mirrors and how much a deep understanding of the code.

**Reply by bigos** [reply]
Wait for a subtle bug introduced by the AI.
